### PR TITLE
features(OSS-868) Implement new driver exceptions hierarchy

### PR DESCRIPTION
**Pull request by fireridlle**
### Notes
[OSS-868 Implement new driver exceptions hierarchy](https://faunadb.atlassian.net/browse/OSS-868)

General idea:
We want to keep everything as simple as possible for the end-user. We can have a base class FaunaException, that consists of the following fields:

httpStatusCode;

code (this one refers to the “code” field from Core);

message (this one refers to the “description” field from Core);

position (array of strings)

### CODE REVIEW

**Review by freels — DISMISSED** [review]

**Comment by faunaee** [comment]
Do we need to use a v5 release branch for this? If this gets merged into main, we won't be able to publish any point releases until API v5 support gets published.

**Comment by fireridlle** [comment]
> Do we need to use a v5 release branch for this? If this gets merged into main, we won't be able to publish any point releases until API v5 support gets published.
yes. have to wait

**Review by freels — APPROVED** [review]

**Review by parkhomenko — APPROVED** [review]

**Comment by parkhomenko** [comment]
@fireridlle Let's resolve the conflicts and merge this PR into `v5` branch.

**Comment by steida** [comment]
Please, do not use exceptions, throw and catch are not type-safe and never will be. Use return object instead. Throw and catch is a well-known anti-pattern. Use something like

```ts
export interface FaunaError {
  type: 'faunaError';
  error:
    | { type: 'BadRequest'; error: errors.BadRequest }
    | { type: 'ClientClosed'; error: errors.ClientClosed }
    | { type: 'InternalError'; error: errors.InternalError }
    | { type: 'InvalidValue'; error: errors.InvalidValue }
    | { type: 'MethodNotAllowed'; error: errors.MethodNotAllowed }
    | { type: 'NotFound'; error: errors.NotFound }
    | { type: 'PermissionDenied'; error: errors.PermissionDenied }
    | { type: 'Unauthorized'; error: errors.Unauthorized }
    | { type: 'UnavailableError'; error: errors.UnavailableError }
    | { type: 'TooManyRequests'; error: errors.TooManyRequests }
    | { type: 'UnknownError'; error: unknown }
    // Not Fauna lib errors.
    | { type: 'abortError' }
    | { type: 'notInitialized' };
}
```

That's how I use Fauna driver to have typed errors.

```ts
return {
        type: 'faunaError',
        error:
          error instanceof errors.BadRequest
            ? { type: 'BadRequest', error }
            : error instanceof errors.ClientClosed
            ? { type: 'ClientClosed', error }
            : error instanceof errors.InternalError
            ? { type: 'InternalError', error }
            : error instanceof errors.InvalidValue
            ? { type: 'InvalidValue', error }
            : error instanceof errors.MethodNotAllowed
            ? { type: 'MethodNotAllowed', error }
            : error instanceof errors.NotFound
            ? { type: 'NotFound', error }
            : error instanceof errors.PermissionDenied
            ? { type: 'PermissionDenied', error }
            : error instanceof errors.Unauthorized
            ? { type: 'Unauthorized', error }
            : error instanceof errors.UnavailableError
            ? { type: 'UnavailableError', error }
            : error instanceof errors.TooManyRequests
            ? { type: 'TooManyRequests', error }
            : error && typeof error === 'object' && 'isAbortError' in error
            ? { type: 'abortError' }
            : { type: 'UnknownError', error },
      };
      ```

https://rlee.dev/practical-guide-to-fp-ts-part-3

**Comment by fireridlle** [comment]
@steida  an you share the link to the article that try/catch is an antipattern, please?

**Comment by steida** [comment]
@fireridlle Sure. For example this one https://dev.to/anthonyjoeseph/either-vs-exception-handling-3jmg

TLDR; The whole problem is that an error in `catch` is untyped and can not be statically typed, because an error can come from any code. If Fauna will decide to add another exception, the TypeScript compiler will not tell me to handle it, and as result, an exception will not be handled until I will read the documentation. This is a well-known problem of languages like C#, Java, etc. Functional languages use Either pattern, which can be typed, therefore we can enforce handling all cases in some switch. It can be easily emulated in TS even without libraries like fp-ts. Check that link.

**Comment by steida** [comment]
@fireridlle What does it mean for a Fauna driver? Don't expose known exceptions, but catch them and return tagged union instead as I do in that error handling example. You can return it in Promise as well. You can attach the call stack, as described in the last link.
